### PR TITLE
Staging smaller chunks of code like git add -p

### PR DESCRIPTION
**Pull request by christianguenter2**
With this commit a new feature which works similar to git add -p
is added. We are now able to patch, which means that we can
commit on line basis. So we are not restricted anymore to commit
only on file basis.

This is a quite huge PR. I did lots of tests, but I think we need more testing and thorough review.

Patching can be started from the stage view.
image

There one can add several lines and stage them.
image

Finally one can commit the patched files as usual.
image

#1736

### CODE REVIEW

File: src/ui/zcl_abapgit_gui_router.clas.abap [hunk 1]
```diff
@@ -22,6 +22,12 @@ CLASS zcl_abapgit_gui_router DEFINITION
       RETURNING VALUE(ri_page) TYPE REF TO zif_abapgit_gui_page
       RAISING   zcx_abapgit_exception.
 
+    METHODS get_page_patch
+      IMPORTING iv_getdata     TYPE clike
+                iv_prev_page   TYPE clike
+      RETURNING VALUE(ri_page) TYPE REF TO zif_abapgit_gui_page
+      RAISING   zcx_abapgit_exception.
+
```
Comment (larshp) [inline src/ui/zcl_abapgit_gui_router.clas.abap]: can this be added in the pages calling the page instead? how data is encoded is handled by the page, so the decoding is also page dependent. Router is central point of merge conflicts

Comment (christianguenter2) [reply src/ui/zcl_abapgit_gui_router.clas.abap]: method is moved

**Comment by larshp** [comment]
might have some time to look into this during the weekend

**Comment by christianguenter2** [comment]
No stress. I had to rebase master because of merge conflicts.

**Comment by larshp** [comment]
Suggestions/follow up:
* Instead of "add remove", just show "add" which then changes to"remove"(might need some spaces not to change the column width, so that the user can click twice the same place to toggle

**Pull request merged**